### PR TITLE
Implement Docs - Sphinx

### PR DESCRIPTION
**Pull request by Skelmis**
### Description
This pull request adds in [Sphinx](https://www.sphinx-doc.org/en/master/index.html) documentation builds for this repo. It also features a guide on how to automate docs creation with `readthedocs`. The relevant `.readthedocs.yaml` is included, so only a repo maintainer should need to set up `readthedocs` on their dashboard before it is fully automated and deployed publically for free.

For readability reasons, among others, the existing docstrings have been modified to follow [Numpy Docstring Standards](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard). This change has no adverse effects, besides making the code easier to read (in my opinion).

### Checklist
- [x] This PR has been tested.

Not sure how to go about testing something unrelated to performance, however, the docs build with no errors. So I am ticking it.

### CODE REVIEW

**Comment by Skelmis** [comment]
Another item to note, it would beneficial to finish documenting files such as `server.py` before merging and setting up `readthedocs`, however, it is not required as `readthedocs` will automatically rebuild the relevant documentation upon new commits. The documentation `.rst` files have been built such that whether or not the documentation is added it will build & work.

File: docs/conf.py
Comment (ShineyDev) [inline docs/conf.py:27]: Pull the version string from [`__init__.py`](https://github.com/Ext-Creators/discord-ext-ipc/blob/master/discord/ext/ipc/__init__.py#L20) not unlike [`setup.py`](https://github.com/Ext-Creators/discord-ext-ipc/blob/master/setup.py#L8-L11).

File: docs/conf.py
Comment (ShineyDev) [inline docs/conf.py:35]: [intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html), [sphinxcontrib_trio](https://github.com/python-trio/sphinxcontrib-trio), and [sphinx_rtd_theme](https://github.com/readthedocs/sphinx_rtd_theme) are also required extensions.

File: docs/conf.py
Comment (ShineyDev) [inline docs/conf.py:38]: This can be removed.

File: docs/conf.py
Comment (ShineyDev) [inline docs/conf.py:43]: This can be removed.

File: docs/conf.py
Comment (ShineyDev) [inline docs/conf.py:56]: This can be removed.

File: docs/conf.py
Comment (ShineyDev) [inline docs/conf.py:1]: This file should also contain the following:
```py
autodoc_typehints = "none"
intersphinx_mapping = {
    "aiohttp": ("https://docs.aiohttp.org/en/stable/", None),
    "python": ("https://docs.python.org/3", None),
}

highlight_language = "python3"
master_doc = "index"
pygments_style = "friendly"
source_suffix = ".rst"
```

File: docs/index.rst
Comment (ShineyDev) [inline docs/index.rst:1]: This comment can be removed.

File: docs/modules/client.rst
Comment (ShineyDev) [inline docs/modules/client.rst:14]: Use the `currentmodule` directive.

**Review by ShineyDev — CHANGES_REQUESTED** [review]
Here's a review of everything that isn't content. On top of fulfilling this review, you should remove the `_build` directory from your PR.

File: .gitignore
Comment (Gobot1234) [reply .gitignore:9]: I'd add docs/_build here similarly to d.py

File: discord/ext/ipc/__init__.py
Comment (Skelmis) [reply discord/ext/ipc/__init__.py:2]: So you want the license showing on all of the docs pages?

File: docs/make.bat
Comment (Skelmis) [reply docs/make.bat:1]: I haven't tested on rtd, but is this not a *required* part of the build process?

File: discord/ext/ipc/__init__.py
Comment (ShineyDev) [reply discord/ext/ipc/__init__.py:2]: I see what you're going for but you should avoid using the `automodule` directive before messying up the file headers.

File: docs/make.bat
Comment (ShineyDev) [reply docs/make.bat:1]: Neither makefile is necessary for Sphinx or RTD.

**Comment by ShineyDev** [comment]
> Not sure how to go about testing something unrelated to performance, however, the docs build with no errors. So I am ticking it.

Provided the existing checks pass and you can build your documentation with `sphinx-build -W -n -a -b html ./docs ./docs/build`, I'd say you pass. I'll push my documentation check later tonight.

File: .gitignore
Comment (Skelmis) [reply .gitignore:9]: Appreciate the heads up. Wasnt aware of ignoring things outside of a `.gitignore` so cheers for that

File: .readthedocs.yaml
Comment (Gobot1234) [inline .readthedocs.yaml:4]: This isn't going to build correctly, you need to add
```yaml
python:
  version: 3.8
  install:
    - method: pip
      path: .
      extra_requirements:
        - docs```I'm pretty sure.

Comment (Skelmis) [reply .readthedocs.yaml:4]: What I have in the files works fine for me on both `rtd` and locally..? Albeit tested against `version: 1` not `version: 2`

File: docs/modules/client.rst
Comment (Skelmis) [reply docs/modules/client.rst:14]: Please advise further? Using that directive I am either running into build errors or a failure to get any documentation outputted?
I haven't used it before, if you could provide an example that'd be beautiful. What I've found online hasn't been helpful so far

File: .readthedocs.yaml
Comment (Gobot1234) [reply .readthedocs.yaml:4]: If it works fine on rtd that's cool then but I would assume it wouldn't due to this being included in their configuration file. https://docs.readthedocs.io/en/stable/config-file/v2.html

Comment (ShineyDev) [reply .readthedocs.yaml:4]: I should have added `extra_requirements` here along with 51316103c25e76c6f2f3f29cf5db9625c65cb697. You [shouldn't be using RTD config v1](https://docs.readthedocs.io/en/stable/config-file/v2.html#migrating-from-v1).

Comment (Skelmis) [reply .readthedocs.yaml:4]: Righto

File: .gitignore
Comment (ShineyDev) [inline .gitignore:7]:
```suggestion
/*.egg-info/

```

File: discord/ext/ipc/__init__.py
Comment (ShineyDev) [inline discord/ext/ipc/__init__.py:6]: Unsure how this happened.

File: docs/conf.py
Comment (ShineyDev) [inline docs/conf.py:9]: These comments can be removed too.

**Review by ShineyDev — CHANGES_REQUESTED** [review]
Some minor cleanup is needed.

**Comment by ShineyDev** [comment]
This is starting to look much better. I'll let you patch up the latest review while the analyze check passes and request that Logan review the documentation content. Thanks for the PR!

File: discord/ext/ipc/__init__.py
Comment (Skelmis) [reply discord/ext/ipc/__init__.py:6]: Sphinx complains about the indentation

File: docs/conf.py
Comment (Gobot1234) [reply docs/conf.py:9]: Isn't this likely to fail cause the package isn't installed at this point?

Comment (Skelmis) [reply docs/conf.py:9]: The sys path is pointed at the directory above, so it should just read the code alongside it. Since 'the repo' is cloned that path exists locally

File: docs/modules/client.rst
Comment (ShineyDev) [reply docs/modules/client.rst:14]: The [`currentmodule` directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html?highlight=currentmodule#directive-py-currentmodule) should exist at the top of a file. The following should be valid:

```rst
.. currentmodule:: discord.ext.ipc.client

.. autoclass:: Client
```

File: discord/ext/ipc/__init__.py
Comment (ShineyDev) [reply discord/ext/ipc/__init__.py:6]: Sphinx shouldn't be reading this docstring at all.

Comment (Skelmis) [reply discord/ext/ipc/__init__.py:6]: ¯\_(ツ)_/¯

**Review by lgaan — CHANGES_REQUESTED** [review]
```py
class Client:
    """Handles webserver side requests to the bot process.
    :param host: The IP or host of the IPC server, defaults to localhost
    :type host: ``str``, optional
    :param port: The port of the IPC server. If not supplied the port will be found automatically, defaults to None
    :type port: ``int``, optional
    :param secret_key: The secret key for your IPC server. Must match the server secret_key or requests will not go ahead, defaults to None
    :type secret_key: ``Union[str, bytes]``, optional
    """
``` should this not be a numpy style docstring?

**Comment by Skelmis** [comment]
> ```python
> class Client:
>     """Handles webserver side requests to the bot process.
>     :param host: The IP or host of the IPC server, defaults to localhost
>     :type host: ``str``, optional
>     :param port: The port of the IPC server. If not supplied the port will be found automatically, defaults to None
>     :type port: ``int``, optional
>     :param secret_key: The secret key for your IPC server. Must match the server secret_key or requests will not go ahead, defaults to None
>     :type secret_key: ``Union[str, bytes]``, optional
>     """
> ```
> 
> should this not be a numpy style docstring?

Your correct. I missed it

File: discord/ext/ipc/__init__.py
Comment (Skelmis) [reply discord/ext/ipc/__init__.py]: Resolved now.

File: docs/modules/server.rst
Comment (ShineyDev) [inline docs/modules/server.rst:26]:
```suggestion
.. autofunction:: route
```

File: docs/modules/server.rst
Comment (ShineyDev) [inline docs/modules/server.rst:27]: Is [IpcServerResponse](https://github.com/Ext-Creators/discord-ext-ipc/blob/master/discord/ext/ipc/server.py#L32) appropriate for documentation?

File: docs/modules/server.rst
Comment (ShineyDev) [inline docs/modules/server.rst:30]: This can be removed with the new content additions.

**Review by ShineyDev — CHANGES_REQUESTED** [review]

File: docs/modules/server.rst
Comment (Skelmis) [reply docs/modules/server.rst:27]: I'll ask Logan if its something used externally, I haven't used it yet myself so I didn't think it was public. Let me ask him and read the code then Ill respond

File: docs/modules/server.rst
Comment (Skelmis) [reply docs/modules/server.rst:30]: wdym

Comment (ShineyDev) [reply docs/modules/server.rst:30]: <sub>This is why showing surrounding lines was a bad idea, GitHub.</sub>

```suggestion

```

Comment (Skelmis) [reply docs/modules/server.rst:30]: > This is why showing surrounding lines was a bad idea, GitHub.

Still lost, but ill remove `:undoc-members:` since Logan *should* have the docstrings done for all relevant methods so there is no need to show ones without docstrings

Comment (ShineyDev) [reply docs/modules/server.rst:30]: That's the idea!

Comment (Skelmis) [reply docs/modules/server.rst:30]: Ah easy

**Review by lgaan — APPROVED** [review]

**Review by lgaan — APPROVED** [review]
lgtm

File: docs/conf.py
Comment (Gobot1234) [inline docs/conf.py:6]: This isn't going to work on rtd. You need to do the same thing that is done in the setup file.
```suggestion
with open("discord/ext/ipc/__init__.py") as stream:
    release = re.search(
        r"^__version__\s*=\s*[\'\"]([^\'\"]*)[\'\"]", stream.read(), re.MULTILINE
    ).group(1)
```

File: discord/ext/ipc/client.py
Comment (ShineyDev) [inline discord/ext/ipc/client.py:23]:
```suggestion
    """
    Handles webserver side requests to the bot process.
```

File: discord/ext/ipc/client.py
Comment (ShineyDev) [inline discord/ext/ipc/client.py:31]:
```suggestion
    secret_key: Union[str, bytes]
```

File: docs/conf.py
Comment (ShineyDev) [inline docs/conf.py:8]: Both [release and version](https://www.sphinx-doc.org/en/master/usage/configuration.html?highlight=conf.py#confval-version) should be defined in this file.

File: docs/modules/examples.rst
Comment (ShineyDev) [inline docs/modules/examples.rst:13]: This needs content.

**Review by ShineyDev — CHANGES_REQUESTED** [review]
Looks mostly good. This is likely to be my final review.

File: docs/modules/server.rst
Comment (Gobot1234) [inline docs/modules/server.rst:20]:
```suggestion
For example, a client request could be {'endpoint': 'get_guild_count', 'headers': {...}}.
```

File: discord/ext/ipc/client.py
Comment (Skelmis) [reply discord/ext/ipc/client.py:31]: This will complain. But I'll figure out something

File: discord/ext/ipc/client.py
Comment (Skelmis) [reply discord/ext/ipc/client.py:23]: This fits the description for a short summary, according to the standards being followed for documentation. But can do

File: docs/conf.py
Comment (Skelmis) [reply docs/conf.py:8]: Aight I'll go pew pew some more logic into it

File: docs/modules/examples.rst
Comment (Skelmis) [reply docs/modules/examples.rst:13]: No offense but not my area to resolve. Up to Logan or y'all to fill them how you see fit

File: discord/ext/ipc/client.py
Comment (ShineyDev) [reply discord/ext/ipc/client.py:23]: I'm sure it does, this is a personal preference.

File: discord/ext/ipc/client.py
Comment (ShineyDev) [reply discord/ext/ipc/client.py:31]: What gives?

File: docs/modules/examples.rst
Comment (ShineyDev) [reply docs/modules/examples.rst:13]: I don't disagree. CC @lgaan

File: discord/ext/ipc/client.py
Comment (Skelmis) [reply discord/ext/ipc/client.py:31]: Can't remember honestly, I'm busy tonight but I'll @ you with it when I work on this

File: docs/conf.py
Comment (Skelmis) [reply docs/conf.py:8]: Erm, how do you want this done? Is there a separate variable somewhere reference if its final, etc?
Or can I just sneakily `version = release`

Comment (ShineyDev) [reply docs/conf.py:8]: The latter.

File: discord/ext/ipc/client.py
Comment (Skelmis) [reply discord/ext/ipc/client.py]: Simply applied your suggestion

**Review by ShineyDev — APPROVED** [review]
LGTM.

**Pull request merged**
